doc/guides/compile.md

```diff
@@ -338,6 +338,47 @@ When debugging mode is enabled
   * Because `-g` flag would be added to `mrbc` runner.
     * You can have better backtrace of mruby scripts with this.
 
+### Additional license terms
+
+The license terms set in the `conf.terms` property are printed as part of
+the `<build-dir>/LICENSE.yml` file.
+This initial value is an empty array (`[]`). If you want to add it, do the
+following:
+
+```ruby
+conf.terms << "path/to/ADDITIONAL-LICENSE"
+```
+
+The file path can be specified as an absolute path as well as a relative
+path based on the build directory.
+Also, you can give not only the file path but also a hash that combines
+the name and the content of the terms.
+
+Objects that can be specified directly in `conf.terms` are `nil`, strings
+that represent paths, array, hash, plus objects that have both `#to_path`
+and `#read` methods (assuming file and pathname).
+
+Inside the array, you can add anything that is given directly to
+`conf.terms`.
+
+In hash, you can specify the object that represents the path (name) as a
+key and the object to be the content as a set.
+The object that represents a path is a string or an object that has a
+`#to_path` method (assuming file or pathname).
+The content object is a string as the terms body, a proc that returns a
+string, or an object with a `#read` method (assuming file, pathname or
+stringio).
+
+The following examples are all correct:
```

Comment:
It feels to me like this API can accept too many kinds of data, is there a simpler way to avoid supporting so many different inputs?

Or what is the use-case for supporting this many patterns?

Reply:
> It feels to me like this API can accept too many kinds of data, is there a simpler way to avoid supporting so many different inputs?
> 
> Or what is the use-case for supporting this many patterns?

Roughly speaking:
  - Specify an object that means the path name in the array
  - For the hash object, specify pair the key that means the path and the objects that are the license terms.
  - The reason the `proc` object can be given as the "value" of a hash object is for lazy evaluation.

I implemented it this way because I didn't want to clearly determine the exact type, but it resulted in a lot of explanation.

If this explanation is inadequate, please ask me again.

Reply:
I think starting with a small API would be better than to introduce many different options we will need to support in the future.

This is just my opinion, but MRuby build config / gem spec are already complex enough that I don't want to introduce even more complexity to their dynamic structure.

Reply:
I can sympathize with your opinion that you don't want to add complexity.
For me, I'm not trying to accommodate many types, I'm just putting together objects that have similar meanings.
This can be broadly divided into "file paths", "clause texts", arrays, and hashes.

This PR implementation requires more branching if nothing is to be done.
Best of all, my eyesight can't distinguish between instances of `Pathname` and `File`, and in some cases `StringIO`. :yum:

These operations are aggregated in the methods related to `Array#fileset_make_under` in `lib/mruby/core_ext.rb`.
  - https://github.com/mruby/mruby/blob/6a4e6608965a1dd3be6b07fb74d6fef43b4f3d39/lib/mruby/core_ext.rb#L32-L80

The trigger is a call to the `fileset_make_under` method in the `tasks/license.rake` file.
  - https://github.com/mruby/mruby/blob/6a4e6608965a1dd3be6b07fb74d6fef43b4f3d39/tasks/license.rake#L71
  - https://github.com/mruby/mruby/blob/6a4e6608965a1dd3be6b07fb74d6fef43b4f3d39/tasks/license.rake#L75
  - https://github.com/mruby/mruby/blob/6a4e6608965a1dd3be6b07fb74d6fef43b4f3d39/tasks/license.rake#L82

Personally, I think it was well organized.
Or is it better to make it even smaller?
However, it is difficult to make a selection in that case.

Reply:
If I were to try and simplify this feature I would _only_ support a string path to file relative to the GEM directory as `spec.terms`. This could be an array of strings for multiple licenses. But if people want to support fetching or injecting licenses from all kinds of places, they should commit them to source in the gem's path.

Combining all LICENSE(s) to a single file in the generated build is a great idea, we're introducing a new field to enable this feature -- so we don't have to worry about breaking existing workflows or compatibility. If we introduce this field with many different options it makes changing later even harder. But adding more options to this field is easier than removing them. This is the point I'm trying to make. :bow:

Reply:
I said it would also support `File` and` StringIO` for implementation reasons.
However, I am planning to change this to the treatment that future operation is not guaranteed and remove it from the document (currently it works).

The composition I want to keep:
  - `[ (string_as_filepath OR pathname), ... ]`
  - `{ (string_as_filepath OR pathname) => (string_as_terms OR pathname OR proc), ... }`

Regarding `Hash` and `Proc`, I think that it is a mechanism that can reduce the burden on the user (and computer).
If I couldn't think of this, I would probably have suggested adding a `LICENSE` file to `mruby-{io,socket,pack,sleep}`.
But I want to avoid making changes to "LICENSE" in other people's projects as much as possible...

> if people want to support fetching or injecting licenses from all kinds of places, they should commit them to source in the gem's path.

If the GEM depends on an external library and there is no single `LICENSE` file, it will have to be maintained up to the external license.
I don't think the GEM author should bear this burden.
Also, if it requires the creation of a temporary file, it is also at the expense of the GEM author.
I think `Hash` and` Proc` can help to avoid these.

> If we introduce this field with many different options it makes changing later even harder.

If possible, could you please tell me what kind of ideas you have?

> Combining all LICENSE(s) to a single file in the generated build is a great idea

The basis for my idea is in Firefox's `about:license`.
I would appreciate it if you could keep it a secret.

Reply:
If I understand correctly you're trying to solve two problems with this PR:

1. Combining LICENSE for all mruby core features, including core mrbgems
2. Enable GEM author to include external license with their packaging

I think this request is reasonable, but I would be weary that this feature may break or change in the future -- causing packages to be actually missing LICENSEs that are required by law which puts us in a tough situation.

But if we delay this decision for now, and we put the work onto the GEM author then it is entirely up to them to solve this problem. For now.

I much prefer enabling authors to solve this problem on their own, than to build a solution ourselves that solves our needs but creates additional maintenance burden and adds more risk to mruby developers.

My reason for pushing back on this so hard is because I want to support a static mruby configuration eventually, and the more complex options we provide the harder it is to support them.

Reply:
> causing packages to be actually missing LICENSEs that are required by law which puts us in a tough situation.

I wrote it in `LICENSE.yml` as "DISCLAIMER FOR THIS FILE" because I was worried about it.

> My reason for pushing back on this so hard is because I want to support a static mruby configuration eventually, and the more complex options we provide the harder it is to support them.

Hmmm, this PR is in complete conflict.
I can't think of a solution at all, so this time I'll just change it to a combination of strings and arrays.

First, I will send PR to add the `LICENSE` file to `mruby-{io,socket,pack,sleep}` and for `mrb_float_read()` in `src/string.c`.
If the PR is unacceptable, we will withdraw this PR as well.

Reply:
@dearblue That sounds very reasonable, thank you for understanding my perspective and working together on this :bow:

Reply:
I updated it. I would appreciate it if you could check it.

Reply:
@dearblue This new version looks much better, and simpler! Great work 🎉